### ISSUE TITLE
model.get() does not clone objects for JSON data-type

### ISSUE THREAD

**Issue by justinkalland** [sequelize/sequelize#10052]
## What are you doing?

```js
const Sequelize = require('sequelize')

const sequelize = new Sequelize({
  dialect: 'sqlite',
  storage: ':memory:'
})

const Album = sequelize.define('User', {
  name: Sequelize.TEXT,
  attributes: Sequelize.JSON
})

Album.sync().then(() => {
  return Album.create({
    name: 'Slow Dance Showdown',
    attributes: {
      artist: 'john smith',
      genre: 'rock'
    }
  })
}).then(album => {
  const dataCopy = album.toJSON()

  dataCopy.attributes.genre = 'classical'
  dataCopy.name = 'Workout Jams'

  console.log('Name compare: ' + (dataCopy.name === album.name)) // works properly
  console.log('Attributes compare: ' + (dataCopy.attributes === album.attributes)) // does not work properly
})
```

## What do you expect to happen?
Using JSON data-type I expect `model.get()` and `model.toJSON()` to return a copy of the object.

## What is actually happening?
Instead of a copy of the object representing the `JSON` you get a reference to the same object that the model uses.

__Dialect:__ mysql / postgres / sqlite
__Dialect version:__ various
__Database version:__ various
__Sequelize version:__ master branch
__Tested with latest release:__ Yes

**Comment by sushantdhiman** [comment]
We need to `cloneDeep` https://github.com/sequelize/sequelize/blob/be958ac1d957cdb0f78e2b07dff8870fdd509456/lib/model.js#L4190

**Comment by SimonSchick** [comment]
@sushantdhiman shouldn't it perform the deep clone here instead? https://github.com/sequelize/sequelize/blob/be958ac1d957cdb0f78e2b07dff8870fdd509456/lib/model.js#L3308 ? then you could probably get rid of the `clone` in the `toJSON` call too.

**Comment by ALiangLiang** [comment]
> @sushantdhiman shouldn't it perform the deep clone here instead?
> 
> [sequelize/lib/model.js](https://github.com/sequelize/sequelize/blob/be958ac1d957cdb0f78e2b07dff8870fdd509456/lib/model.js#L3308)
> 
> Line 3308 in [be958ac](/sequelize/sequelize/commit/be958ac1d957cdb0f78e2b07dff8870fdd509456)
> 
>  if (Array.isArray(this.dataValues[key])) { 
> ? then you could probably get rid of the `clone` in the `toJSON` call too.

It's only deep clone `Array` data.

**Comment by justinkalland** [comment]
@sushantdhiman That is an easy fix, and I will submit a pull request. I just wasn't sure what the intended behavior of `.get()` without a key was supposed to be. Do you believe only `model.toJSON()` should return a clone, but not `model.get()`? It seems in some cases `.get` returns a clone and some cases `this.dataValues`.

**Comment by SimonSchick** [comment]
@sushantdhiman I wasn't talking about the exact line but the general location, you'd have to check if `typeof this.dataValues[key] === 'object'` before returning the data, avoid having to deep clone the entire object.

**Comment by sushantdhiman** [comment]
> Do you believe only model.toJSON() should return a clone, but not model.get()? It seems in some cases .get returns a clone and some cases this.dataValues. This is why I submitted an issue instead of just a pull request with cloneDeep.

I personally believe they both (and other cases) all should return a deep cloned value. We never want mutations from `get()` or `toJSON()` to affect our next save or update call.

As for its implications I am not so sure

1. There is a clone option for `get`, but I don't see it used any where
2. Performance impact, I see internal codebase is using mostly key based case which should not be affected. This is just a concern, any incorrect behaviour should be correct regardless of performance impact.

This code is so old I am not sure, @justinkalland I say let's try a PR which iron out these cases so we return deep cloned value for all cases. As @SimonSchick suggested we can use selective cases to avoid cloning whole objects

**Comment by justinkalland** [comment]
I agree @sushantdhiman ,`get()` should always returned a deep clone value, and `toJSON()` should just alias `get()` with no key (as it is documented as doing). But I suspect that the inconsistent behavior of `get()` may be relied upon in sequelize.

**Comment by justinkalland** [comment]
I worked on a fix for this but changing `get()` gets complicated. It has a default option `options.plain=false` and so by default it returns references to included Model instances (rather than "clones" of their data). This behavior is in use many places. If this was a new library I would argue to remove `options.plain` and always return cloned. But with the age of Sequelize perhaps just changing the toJSON to cloneDeep would be an acceptable solution.

**Comment by sushantdhiman** [comment]
@justinkalland Yeah deep clone for toJSON is good enough
